driver/gather_driver.hpp

```diff
@@ -0,0 +1,378 @@
+    miopenTensorDescriptor_t indicesTensor;
+    miopenTensorDescriptor_t paramGradTensor;
+
+    std::unique_ptr<GPUMem> outputGrad_dev;
+    std::unique_ptr<GPUMem> indices_dev;
+    std::unique_ptr<GPUMem> paramGrad_dev;
+
+    std::vector<Tgpu> outGrad;
+    std::vector<Tindex> indices;
+    std::vector<Tgpu> paramGrad;
+
+    std::vector<Tref> paramGradHost;
+
+    miopenGatherMode_t mode;
+    uint32_t dim;
+    uint32_t batch_dims;
+};
+
+template <typename Tgpu, typename Tref, typename Tindex>
+int GatherDriver<Tgpu, Tref, Tindex>::ParseCmdLineArgs(int argc, char* argv[])
+{
+    inflags.Parse(argc, argv);
+
+    forw = inflags.GetValueInt("forw");
+    MIOPEN_THROW_IF(forw != 0, "Incorrect Forward Mode");
+
+    if(inflags.GetValueInt("time") == 1)
+    {
+        miopenEnableProfiling(GetHandle(), true);
+    }
+
+    dim        = inflags.GetValueInt("dim");
+    batch_dims = inflags.GetValueInt("batch_dims");
+    if(inflags.GetValueStr("mode") == "gather")
+    {
+        mode = MIOPEN_GATHER;
+    }
+    else if(inflags.GetValueStr("mode") == "gatherv2")
+    {
+        mode = MIOPEN_GATHER_V2;
+    }
+    else if(inflags.GetValueStr("mode") == "gathernd")
+    {
+        mode = MIOPEN_GATHER_ND;
+    }
+    else
+    {
+        MIOPEN_THROW("Incorrect Gather Mode");
+    }
+
+    miopenCreateGatherDescriptor(&gatherDesc);
+    miopenSetGatherDescriptor(gatherDesc, mode, dim, batch_dims);
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref, typename Tindex>
+int GatherDriver<Tgpu, Tref, Tindex>::GetandSetData()
+{
+    std::vector<int> paramGrad_len = inflags.GetValueTensor("param_grad_shape").lengths;
+    SetTensorNd(paramGradTensor, paramGrad_len, data_type);
+
+    std::vector<int> indices_len = inflags.GetValueTensor("indices_shape").lengths;
+    SetTensorNd(indicesTensor, indices_len, index_data_type);
+
+    std::vector<int> outGrad_len;
+
+    // output shape = param[:axis] + indice[batch_dim:] + param[axis + 1:]
+    if(mode == MIOPEN_GATHER_V2)
+    {
+        for(uint32_t i = 0; i < dim; i++)
+        {
+            outGrad_len.push_back(paramGrad_len[i]);
+        }
+
+        for(uint32_t i = batch_dims; i < indices_len.size(); i++)
+        {
+            outGrad_len.push_back(indices_len[i]);
+        }
+
+        for(uint32_t i = dim + 1; i < paramGrad_len.size(); i++)
+        {
+            outGrad_len.push_back(paramGrad_len[i]);
+        }
+    }
+    else
+    {
+        return miopenStatusNotImplemented;
+    }
+
+    SetTensorNd(outputGradTensor, outGrad_len, data_type);
+
+    return miopenStatusSuccess;
+}
+
+template <typename Tgpu, typename Tref, typename Tindex>
+int GatherDriver<Tgpu, Tref, Tindex>::AddCmdLineArgs()
+{
+    inflags.AddInputFlag("forw",
+                         'F',
+                         "0",
+                         "Run only Forward (1) or Run both Forward and Backward (0) (Default = 0)",
+                         "int");
```

Comment:
You raise error when -F != 0, so this description is confusing

Reply:
updated it.